File: laravel-backend/app/Services/InvoiceService.php
<?php

namespace App\Services;

use Illuminate\Support\Facades\DB;
use Illuminate\Support\Str;

class InvoiceService
{
    protected SMSService $smsService;
    protected PaymentGatewayService $paymentService;

    public function __construct(SMSService $smsService, PaymentGatewayService $paymentService)
    {
        $this->smsService = $smsService;
        $this->paymentService = $paymentService;
    }

    /**
     * Generate unique invoice number
     */
    public function generateInvoiceNumber(): string
    {
        $prefix = $this->getSetting('invoice_prefix', 'INV');
        $year = date('Y');
        
        $lastInvoice = DB::connection('tenant')
            ->table('invoices')
            ->where('invoice_number', 'like', "{$prefix}-{$year}-%")
            ->orderBy('id', 'desc')
            ->first();

        if ($lastInvoice) {
            $lastNumber = (int) Str::afterLast($lastInvoice->invoice_number, '-');
            $newNumber = $lastNumber + 1;
        } else {
            $newNumber = 1;
        }

        return sprintf('%s-%s-%06d', $prefix, $year, $newNumber);
    }

    /**
     * Create a new invoice
     */
    public function createInvoice(array $data): array
    {
        $business = DB::connection('tenant')
            ->table('businesses')
            ->where('id', $data['business_id'])
            ->first();

        if (!$business) {
            return ['success' => false, 'message' => 'Business not found'];
        }

        $revenueItem = DB::connection('tenant')
            ->table('revenue_items')
            ->where('id', $data['revenue_item_id'])
            ->first();

        if (!$revenueItem) {
            return ['success' => false, 'message' => 'Revenue item not found'];
        }

        // Calculate amount based on tariff rules
        $amount = $this->calculateAmount($revenueItem->id, $business);
        if (isset($data['amount'])) {
            $amount = $data['amount']; // Allow override
        }

        // Calculate revenue share
        $tenant = app('current_tenant');
        $platformFee = $this->calculatePlatformFee($amount, $tenant);
        $netLgaAmount = $amount - $platformFee;

        $invoiceNumber = $this->generateInvoiceNumber();
        $dueDate = $data['due_date'] ?? date('Y-m-d', strtotime('+' . $this->getSetting('default_due_days', 30) . ' days'));

        $invoiceId = DB::connection('tenant')->table('invoices')->insertGetId([
            'invoice_number' => $invoiceNumber,
            'business_id' => $business->id,
            'revenue_item_id' => $revenueItem->id,
            'created_by' => $data['created_by'] ?? null,
            'subtotal' => $amount,
            'tax_amount' => 0,
            'total_amount' => $amount,
            'amount_paid' => 0,
            'balance' => $amount,
            'platform_fee' => $platformFee,
            'net_lga_amount' => $netLgaAmount,
            'description' => $data['description'] ?? $revenueItem->name,
            'issue_date' => $data['issue_date'] ?? date('Y-m-d'),
            'due_date' => $dueDate,
            'period' => $data['period'] ?? $revenueItem->frequency,
            'fiscal_year' => date('Y'),
            'status' => 'draft',
            'delivery_status' => 'not_sent',
            'created_at' => now(),
            'updated_at' => now(),
        ]);

        // Create invoice item
        DB::connection('tenant')->table('invoice_items')->insert([
            'invoice_id' => $invoiceId,
            'revenue_item_id' => $revenueItem->id,
            'description' => $revenueItem->name,
            'quantity' => 1,
            'unit_price' => $amount,
            'amount' => $amount,
            'created_at' => now(),
            'updated_at' => now(),
        ]);

        // Log audit
        $this->logAudit('create', 'invoices', 'Invoice', $invoiceId, [
            'invoice_number' => $invoiceNumber,
            'business_name' => $business->name,
            'amount' => $amount,
        ]);

        return [
            'success' => true,
            'invoice_id' => $invoiceId,
            'invoice_number' => $invoiceNumber,
        ];
    }

    /**
     * Issue an invoice (change status from draft to issued)
     */
    public function issueInvoice(int $invoiceId, bool $sendSMS = true): array
    {
        $invoice = DB::connection('tenant')
            ->table('invoices')
            ->where('id', $invoiceId)
            ->first();

        if (!$invoice) {
            return ['success' => false, 'message' => 'Invoice not found'];
        }

        // Generate payment link and QR code
        $paymentLink = $this->generatePaymentLink($invoice);

        DB::connection('tenant')->table('invoices')
            ->where('id', $invoiceId)
            ->update([
                'status' => 'issued',
                'payment_link' => $paymentLink,
                'updated_at' => now(),
            ]);

        // Send SMS notification
        if ($sendSMS && $this->getSetting('auto_send_invoice_sms', true)) {
            $this->sendInvoiceIssuedSMS($invoiceId);
        }

        $this->logAudit('issue', 'invoices', 'Invoice', $invoiceId, [
            'invoice_number' => $invoice->invoice_number,
        ]);

        return ['success' => true, 'message' => 'Invoice issued successfully'];
    }

    /**
     * Bulk generate invoices
     */
    public function bulkGenerateInvoices(array $criteria, int $revenueItemId, array $options = []): array
    {
        $query = DB::connection('tenant')
            ->table('businesses')
            ->where('status', 'active');

        if (!empty($criteria['ward_id'])) {
            $query->where('ward_id', $criteria['ward_id']);
        }

        if (!empty($criteria['department_id'])) {
            $query->where('department_id', $criteria['department_id']);
        }

        if (!empty($criteria['business_size'])) {
            $query->where('business_size', $criteria['business_size']);
        }

        if (!empty($criteria['business_category'])) {
            $query->where('business_category', $criteria['business_category']);
        }

        $businesses = $query->get();
        $results = [];

        foreach ($businesses as $business) {
            // Check if invoice already exists for this period
            $existingInvoice = DB::connection('tenant')
                ->table('invoices')
                ->where('business_id', $business->id)
                ->where('revenue_item_id', $revenueItemId)
                ->where('fiscal_year', date('Y'))
                ->whereNotIn('status', ['cancelled'])
                ->first();

            if ($existingInvoice && !($options['force_create'] ?? false)) {
                $results[$business->id] = [
                    'success' => false,
                    'message' => 'Invoice already exists',
                    'invoice_id' => $existingInvoice->id,
                ];
                continue;
            }

            $result = $this->createInvoice([
                'business_id' => $business->id,
                'revenue_item_id' => $revenueItemId,
                'created_by' => $options['created_by'] ?? null,
                'due_date' => $options['due_date'] ?? null,
            ]);

            $results[$business->id] = $result;

            // Auto-issue if requested
            if ($result['success'] && ($options['auto_issue'] ?? false)) {
                $this->issueInvoice($result['invoice_id'], $options['send_sms'] ?? true);
            }
        }

        return [
            'success' => true,
            'total_processed' => count($businesses),
            'created' => count(array_filter($results, fn($r) => $r['success'])),
            'failed' => count(array_filter($results, fn($r) => !$r['success'])),
            'results' => $results,
        ];
    }

    /**
     * Record a payment for an invoice
     */
    public function recordPayment(int $invoiceId, array $data): array
    {
        $invoice = DB::connection('tenant')
            ->table('invoices')
            ->where('id', $invoiceId)
            ->first();

        if (!$invoice) {
            return ['success' => false, 'message' => 'Invoice not found'];
        }

        $amount = $data['amount'];
        if ($amount > $invoice->balance) {
            return ['success' => false, 'message' => 'Payment amount exceeds balance'];
        }

        $paymentId = DB::connection('tenant')->table('invoice_payments')->insertGetId([
            'invoice_id' => $invoiceId,
            'amount' => $amount,
            'payment_method' => $data['payment_method'] ?? 'cash',
            'payment_reference' => $data['payment_reference'] ?? $this->generatePaymentReference(),
            'status' => 'success',
            'received_by' => $data['received_by'] ?? null,
            'notes' => $data['notes'] ?? null,
            'paid_at' => now(),
            'created_at' => now(),
            'updated_at' => now(),
        ]);

        $newAmountPaid = $invoice->amount_paid + $amount;
        $newBalance = $invoice->total_amount - $newAmountPaid;
        $newStatus = $newBalance <= 0 ? 'paid' : 'partial';

        DB::connection('tenant')->table('invoices')
            ->where('id', $invoiceId)
            ->update([
                'amount_paid' => $newAmountPaid,
                'balance' => max(0, $newBalance),
                'status' => $newStatus,
                'updated_at' => now(),
            ]);

        // Update defaulter status
        if ($newStatus === 'paid') {
            DB::connection('tenant')->table('defaulters')
                ->where('invoice_id', $invoiceId)
                ->update(['status' => 'resolved', 'updated_at' => now()]);
        }

        // Send payment receipt SMS
        $this->sendPaymentReceiptSMS($invoiceId, $amount, $data['payment_reference'] ?? '');

        $this->logAudit('payment_recorded', 'invoices', 'Invoice', $invoiceId, [
            'amount' => $amount,
            'payment_method' => $data['payment_method'] ?? 'cash',
        ]);

        return [
            'success' => true,
            'payment_id' => $paymentId,
            'new_balance' => $newBalance,
            'invoice_status' => $newStatus,
        ];
    }

    /**
     * Calculate amount based on tariff rules
     */
    protected function calculateAmount(int $revenueItemId, object $business): float
    {
        // First check for specific tariff rule
        $tariff = DB::connection('tenant')
            ->table('tariff_rules')
            ->where('revenue_item_id', $revenueItemId)
            ->where('is_active', true)
            ->where(function ($query) use ($business) {
                $query->where(function ($q) use ($business) {
                    $q->where('business_size', $business->business_size)
                      ->where('business_category', $business->business_category);
                })->orWhere(function ($q) use ($business) {
                    $q->where('business_size', $business->business_size)
                      ->whereNull('business_category');
                })->orWhere(function ($q) use ($business) {
                    $q->whereNull('business_size')
                      ->where('business_category', $business->business_category);
                });
            })
            ->orderByRaw('CASE WHEN business_size IS NOT NULL AND business_category IS NOT NULL THEN 1 
                              WHEN business_size IS NOT NULL THEN 2 
                              WHEN business_category IS NOT NULL THEN 3 
                              ELSE 4 END')
            ->first();

        if ($tariff) {
            return $tariff->amount;
        }

        // Fall back to base amount
        $revenueItem = DB::connection('tenant')
            ->table('revenue_items')
            ->where('id', $revenueItemId)
            ->first();

        return $revenueItem?->base_amount ?? 0;
    }

    /**
     * Calculate platform fee
     */
    protected function calculatePlatformFee(float $amount, ?object $tenant): float
    {
        if (!$tenant) return 0;

        if ($tenant->revenue_share_model === 'percentage') {
            return round($amount * ($tenant->share_value / 100), 2);
        }

        return $tenant->share_value;
    }

    protected function generatePaymentLink(object $invoice): string
    {
        $baseUrl = config('app.url');
        return "{$baseUrl}/pay/{$invoice->invoice_number}";
    }

    protected function generatePaymentReference(): string
    {
        return 'PAY-' . strtoupper(Str::random(10));
    }

    protected function sendInvoiceIssuedSMS(int $invoiceId): void
    {
        $invoice = DB::connection('tenant')
            ->table('invoices')
            ->join('businesses', 'invoices.business_id', '=', 'businesses.id')
            ->where('invoices.id', $invoiceId)
            ->select('invoices.*', 'businesses.owner_phone', 'businesses.name as business_name')
            ->first();

        if (!$invoice) return;

        $tenant = app('current_tenant');
        
        $this->smsService->sendTemplate('invoice_issued', $invoice->owner_phone, [
            'business_name' => $invoice->business_name,
            'invoice_no' => $invoice->invoice_number,
            'amount' => number_format($invoice->total_amount, 2),
            'due_date' => date('d/m/Y', strtotime($invoice->due_date)),
            'payment_link' => $invoice->payment_link ?? '',
            'tenant_name' => $tenant->name ?? 'LGA',
        ], 'Invoice', $invoiceId);
    }

    protected function sendPaymentReceiptSMS(int $invoiceId, float $amount, string $reference): void
    {
        $invoice = DB::connection('tenant')
            ->table('invoices')
            ->join('businesses', 'invoices.business_id', '=', 'businesses.id')
            ->where('invoices.id', $invoiceId)
            ->select('invoices.*', 'businesses.owner_phone', 'businesses.name as business_name')
            ->first();

        if (!$invoice) return;

        $tenant = app('current_tenant');
        
        $this->smsService->sendTemplate('invoice_paid', $invoice->owner_phone, [
            'business_name' => $invoice->business_name,
            'invoice_no' => $invoice->invoice_number,
            'amount' => number_format($amount, 2),
            'reference' => $reference,
            'tenant_name' => $tenant->name ?? 'LGA',
        ], 'Invoice', $invoiceId);
    }

    protected function getSetting(string $key, $default = null)
    {
        $setting = DB::connection('tenant')
            ->table('tenant_settings')
            ->where('key', $key)
            ->first();

        if (!$setting) return $default;

        return match($setting->type) {
            'boolean' => filter_var($setting->value, FILTER_VALIDATE_BOOLEAN),
            'integer' => (int) $setting->value,
            default => $setting->value,
        };
    }

    protected function logAudit(string $action, string $module, string $entityType, int $entityId, array $details): void
    {
        DB::connection('tenant')->table('tenant_audit_logs')->insert([
            'user_id' => auth()->id(),
            'user_type' => 'tenant_user',
            'action' => $action,
            'module' => $module,
            'entity_type' => $entityType,
            'entity_id' => $entityId,
            'details' => json_encode($details),
            'ip_address' => request()->ip(),
            'user_agent' => request()->userAgent(),
            'created_at' => now(),
            'updated_at' => now(),
        ]);
    }
}

File: backend/app/services/invoice_service.py
"""
Invoice Service - Manages invoice lifecycle and payments
"""
from typing import Optional, List
from datetime import datetime, timezone, date, timedelta
from app.database import db_manager
from app.models.tenant import (
    Invoice, InvoiceCreate, InvoiceItem, InvoicePayment, 
    InvoiceStatus, Defaulter, Business
)
import uuid

class InvoiceService:
    def __init__(self, tenant_slug: str):
        self.tenant_slug = tenant_slug
        self.db = db_manager.get_tenant_db(tenant_slug)
    
    async def generate_invoice_number(self) -> str:
        """Generate unique invoice number"""
        year = datetime.now().year
        # Count existing invoices for this year
        count = await self.db.invoices.count_documents({
            "invoice_number": {"$regex": f"^INV-{year}-"}
        })
        return f"INV-{year}-{str(count + 1).zfill(6)}"
    
    async def create_invoice(self, data: InvoiceCreate, created_by: str) -> Invoice:
        """Create a new invoice"""
        invoice_number = await self.generate_invoice_number()
        
        # Calculate totals
        subtotal = sum(item.amount for item in data.items)
        tax_amount = 0  # Can be configured per tenant
        total_amount = subtotal + tax_amount
        
        invoice = Invoice(
            id=str(uuid.uuid4()),
            invoice_number=invoice_number,
            business_id=data.business_id,
            ward_id=data.ward_id,
            status=InvoiceStatus.DRAFT,
            subtotal=subtotal,
            tax_amount=tax_amount,
            total_amount=total_amount,
            amount_paid=0,
            balance=total_amount,
            due_date=data.due_date,
            notes=data.notes,
            created_at=datetime.now(timezone.utc),
            created_by=created_by,
        )
        
        # Save invoice
        invoice_dict = invoice.model_dump()
        invoice_dict['created_at'] = invoice_dict['created_at'].isoformat()
        invoice_dict['due_date'] = invoice_dict['due_date'].isoformat()
        await self.db.invoices.insert_one(invoice_dict)
        
        # Save invoice items
        for item_data in data.items:
            item = InvoiceItem(
                id=str(uuid.uuid4()),
                invoice_id=invoice.id,
                revenue_item_id=item_data.revenue_item_id,
                description=item_data.description,
                quantity=item_data.quantity,
                unit_price=item_data.unit_price,
                amount=item_data.amount,
            )
            await self.db.invoice_items.insert_one(item.model_dump())
        
        return invoice
    
    async def issue_invoice(self, invoice_id: str, issued_by: str) -> Optional[Invoice]:
        """Issue an invoice (change from draft to issued)"""
        await self.db.invoices.update_one(
            {"id": invoice_id, "status": InvoiceStatus.DRAFT},
            {"$set": {
                "status": InvoiceStatus.ISSUED,
                "issued_at": datetime.now(timezone.utc).isoformat(),
            }}
        )
        return await self.get_invoice(invoice_id)
    
    async def record_payment(self, invoice_id: str, amount: float, 
                            payment_method: str, reference: str,
                            received_by: str, gateway: str = None,
                            notes: str = None) -> Optional[InvoicePayment]:
        """Record a payment against an invoice"""
        invoice = await self.get_invoice(invoice_id)
        if not invoice:
            return None
        
        # Create payment record
        payment = InvoicePayment(
            id=str(uuid.uuid4()),
            invoice_id=invoice_id,
            amount=amount,
            payment_method=payment_method,
            reference=reference,
            gateway=gateway,
            received_by=received_by,
            notes=notes,
            created_at=datetime.now(timezone.utc),
        )
        payment_dict = payment.model_dump()
        payment_dict['created_at'] = payment_dict['created_at'].isoformat()
        await self.db.invoice_payments.insert_one(payment_dict)
        
        # Update invoice amounts
        new_amount_paid = invoice.amount_paid + amount
        new_balance = invoice.total_amount - new_amount_paid
        
        # Determine new status
        if new_balance <= 0:
            new_status = InvoiceStatus.PAID
        elif new_amount_paid > 0:
            new_status = InvoiceStatus.PARTIALLY_PAID
        else:
            new_status = invoice.status
        
        update_data = {
            "amount_paid": new_amount_paid,
            "balance": max(0, new_balance),
            "status": new_status,
        }
        if new_status == InvoiceStatus.PAID:
            update_data["paid_at"] = datetime.now(timezone.utc).isoformat()
        
        await self.db.invoices.update_one(
            {"id": invoice_id},
            {"$set": update_data}
        )
        
        # Update business balance
        await self.db.businesses.update_one(
            {"id": invoice.business_id},
            {"$inc": {"total_paid": amount, "balance": -amount}}
        )
        
        # Remove from defaulters if paid
        if new_status == InvoiceStatus.PAID:
            await self.db.defaulters.delete_one({"invoice_id": invoice_id})
        
        return payment
    
    async def get_invoice(self, invoice_id: str) -> Optional[Invoice]:
        """Get invoice by ID"""
        doc = await self.db.invoices.find_one({"id": invoice_id}, {"_id": 0})
        if doc:
            # Convert string dates back
            if isinstance(doc.get('due_date'), str):
                doc['due_date'] = date.fromisoformat(doc['due_date'])
            if isinstance(doc.get('created_at'), str):
                doc['created_at'] = datetime.fromisoformat(doc['created_at'])
            return Invoice(**doc)
        return None
    
    async def get_invoice_items(self, invoice_id: str) -> List[InvoiceItem]:
        """Get items for an invoice"""
        cursor = self.db.invoice_items.find({"invoice_id": invoice_id}, {"_id": 0})
        items = []
        async for doc in cursor:
            items.append(InvoiceItem(**doc))
        return items
    
    async def get_invoice_payments(self, invoice_id: str) -> List[InvoicePayment]:
        """Get payments for an invoice"""
        cursor = self.db.invoice_payments.find({"invoice_id": invoice_id}, {"_id": 0})
        payments = []
        async for doc in cursor:
            if isinstance(doc.get('created_at'), str):
                doc['created_at'] = datetime.fromisoformat(doc['created_at'])
            payments.append(InvoicePayment(**doc))
        return payments
    
    async def list_invoices(self, skip: int = 0, limit: int = 100,
                           business_id: str = None, status: InvoiceStatus = None,
                           ward_id: str = None) -> List[Invoice]:
        """List invoices with filters"""
        query = {}
        if business_id:
            query["business_id"] = business_id
        if status:
            query["status"] = status
        if ward_id:
            query["ward_id"] = ward_id
        
        cursor = self.db.invoices.find(query, {"_id": 0}).skip(skip).limit(limit).sort("created_at", -1)
        invoices = []
        async for doc in cursor:
            if isinstance(doc.get('due_date'), str):
                doc['due_date'] = date.fromisoformat(doc['due_date'])
            if isinstance(doc.get('created_at'), str):
                doc['created_at'] = datetime.fromisoformat(doc['created_at'])
            invoices.append(Invoice(**doc))
        return invoices
    
    async def bulk_generate_invoices(self, business_ids: List[str], 
                                    items: List[dict], due_date: date,
                                    created_by: str) -> List[Invoice]:
        """Generate invoices in bulk for multiple businesses"""
        invoices = []
        for business_id in business_ids:
            # Get business ward
            business = await self.db.businesses.find_one({"id": business_id}, {"_id": 0})
            if not business:
                continue
            
            from app.models.tenant import InvoiceItemBase
            invoice_items = [InvoiceItemBase(**item) for item in items]
            
            invoice_data = InvoiceCreate(
                business_id=business_id,
                ward_id=business.get("ward_id"),
                due_date=due_date,
                items=invoice_items,
            )
            invoice = await self.create_invoice(invoice_data, created_by)
            invoices.append(invoice)
        
        return invoices
    
    async def check_and_mark_overdue(self):
        """Check for overdue invoices and mark them"""
        today = date.today().isoformat()
        
        # Find invoices that are past due date and not paid
        await self.db.invoices.update_many(
            {
                "status": {"$in": [InvoiceStatus.ISSUED, InvoiceStatus.PARTIALLY_PAID]},
                "due_date": {"$lt": today}
            },
            {"$set": {"status": InvoiceStatus.OVERDUE}}
        )
        
        # Create/update defaulter records
        cursor = self.db.invoices.find({"status": InvoiceStatus.OVERDUE}, {"_id": 0})
        async for invoice in cursor:
            due_date = date.fromisoformat(invoice['due_date']) if isinstance(invoice['due_date'], str) else invoice['due_date']
            days_overdue = (date.today() - due_date).days
            
            existing = await self.db.defaulters.find_one({"invoice_id": invoice['id']})
            if existing:
                await self.db.defaulters.update_one(
                    {"invoice_id": invoice['id']},
                    {"$set": {"days_overdue": days_overdue, "amount_owed": invoice['balance']}}
                )
            else:
                defaulter = Defaulter(
                    id=str(uuid.uuid4()),
                    business_id=invoice['business_id'],
                    invoice_id=invoice['id'],
                    amount_owed=invoice['balance'],
                    days_overdue=days_overdue,
                    created_at=datetime.now(timezone.utc),
                )
                defaulter_dict = defaulter.model_dump()
                defaulter_dict['created_at'] = defaulter_dict['created_at'].isoformat()
                await self.db.defaulters.insert_one(defaulter_dict)
    
    async def get_invoice_stats(self) -> dict:
        """Get invoice statistics"""
        pipeline = [
            {"$group": {
                "_id": "$status",
                "count": {"$sum": 1},
                "total": {"$sum": "$total_amount"},
                "paid": {"$sum": "$amount_paid"},
                "balance": {"$sum": "$balance"}
            }}
        ]
        
        stats = {
            "total_invoices": 0,
            "total_amount": 0,
            "total_paid": 0,
            "total_balance": 0,
            "by_status": {}
        }
        
        async for result in self.db.invoices.aggregate(pipeline):
            status = result['_id']
            stats['by_status'][status] = {
                "count": result['count'],
                "total": result['total'],
                "paid": result['paid'],
                "balance": result['balance']
            }
            stats['total_invoices'] += result['count']
            stats['total_amount'] += result['total']
            stats['total_paid'] += result['paid']
            stats['total_balance'] += result['balance']
        
        return stats
